sam2 block

hansent opened the pull request:
# Description
WIP

List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] This change requires a documentation update

## How has this change been tested, please provide a testcase or example of how you tested the change?

YOUR_ANSWER

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

PawelPeczek-Roboflow commented on .release/pypi/inference.core.setup.py:
```diff
@@ -65,7 +65,6 @@ def read_requirements(path):
         "hosted": read_requirements("requirements/requirements.hosted.txt"),
         "http": read_requirements("requirements/requirements.http.txt"),
         "sam": read_requirements("requirements/requirements.sam.txt"),
-        "sam2": read_requirements("requirements/requirements.sam2.txt"),
```
was that extras ever released?

hansent replied:
Yes as part of last release I think.  Probably not very many installs if any; but maybe its possible to leave it as empty?

PawelPeczek-Roboflow replied:
👍

PawelPeczek-Roboflow replied:
lets remove then

probicheaux reviewed (approved):
lgtm, thomas and I paired on this PR

PawelPeczek-Roboflow commented on inference/core/entities/requests/sam2.py:
```diff
@@ -56,6 +56,61 @@ class Sam2EmbeddingRequest(Sam2InferenceRequest):
 
 
+class Box(BaseModel):
+    x: float = Field()
```
I believe that empty `Field(...)` object is not needed

PawelPeczek-Roboflow commented on inference/core/entities/requests/sam2.py:
```diff
@@ -56,6 +56,61 @@ class Sam2EmbeddingRequest(Sam2InferenceRequest):
+            if isinstance(v, list):
+                lengths.add(len(v))
+
+        assert len(lengths) in [0, 1], "All prompts must have the same number of points"
```
please change assert into error

PawelPeczek-Roboflow replied:
Please provide explanation for that assertion

PawelPeczek-Roboflow commented on inference/core/entities/requests/sam2.py:
```diff
@@ -56,6 +56,61 @@ class Sam2EmbeddingRequest(Sam2InferenceRequest):
+
+
+class Sam2PromptSet(BaseModel):
+    prompts: Optional[List[Sam2Prompt]] = Field(default=None)
```
Q: What does empty prompts set semantically mean?

PawelPeczek-Roboflow replied:
lets avoid mutations for payload entities if possible

probicheaux replied:
>Q: What does empty prompts set semantically mean?

If you don't prompt the model, it'll just return the mask of the object most in the foreground

PawelPeczek-Roboflow replied:
👍

PawelPeczek-Roboflow commented on inference/core/entities/requests/sam2.py:
```diff
@@ -80,13 +135,19 @@ class Sam2SegmentationRequest(Sam2InferenceRequest):
-        default=[[0.0, 0.0]],
-        examples=[[[10.0, 10.0]]],
-        description="The coordinates of the interactive points used during decoding. Each point (x,y pair) corresponds to a label in point_labels.",
+    prompts: Sam2PromptSet = Field(default=Sam2PromptSet(prompts=None))
```
description is missing

hansent replied:
added

PawelPeczek-Roboflow commented on inference/core/env.py:
```diff
@@ -106,7 +106,7 @@
 # Flag to enable SAM core model, default is True
 CORE_MODEL_SAM_ENABLED = str2bool(os.getenv("CORE_MODEL_SAM_ENABLED", True))
-CORE_MODEL_SAM2_ENABLED = str2bool(os.getenv("CORE_MODEL_SAM2_ENABLED", False))
+CORE_MODEL_SAM2_ENABLED = str2bool(os.getenv("CORE_MODEL_SAM2_ENABLED", True))
```
do we want to support SAM2 at hosted platform within all other core models?

probicheaux replied:
That's the goal

PawelPeczek-Roboflow replied:
then we must:
* add remote execution into workflows block
* add functional tests for the suite of platform tests

hansent replied:
We changed plans on this. For now (maybe forever) we want to ship SAM2 with support for GPU mainly (OK to allow enabling on CPU via env var).  But for hosted deployments we will not support SAM2 on lambda.  Want to use it to promote new dedicated inference server feature Shuyang has built.  Working on UI for that in workflows next week (will label certain blocks as needing GPU and suggest / let you spin up dedicated dev GPU for building in workflows UI)

hansent replied:
We can still decide later to support it on hosted if we do more benchmarking and implement remote execution for block.

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+)
+
+LONG_DESCRIPTION = """
+Run Segment Anything 2 Model
```
please provide more description about the model for people less advanced technologically

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+        },
+        protected_namespaces=(),
+    )
+    type: Literal["roboflow_core/segment_anything@v1", "SegmentAnything2Model"]
```
No need for double alias

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+        default=None,
+    )
+
+    sam2_model: Union[
```
I vote for name `version` or `model_version` or `model_variant` - we use `version` for YoloWorld

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+    sam2_model: Union[
+        WorkflowParameterSelector(kind=[STRING_KIND]),
+        Literal["hiera_large", "hiera_small", "hiera_tiny", "hiera_b_plus"],
+    ] = Field(
```
let's ask @SkalskiP what does he recommend as default

PawelPeczek-Roboflow replied:
thanks, lets apply then

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+    ) -> BlockResult:
+        if self._step_execution_mode is StepExecutionMode.LOCAL:
+            return self.run_locally(images=images, sam2_model=sam2_model, boxes=boxes)
+        elif self._step_execution_mode is StepExecutionMode.REMOTE:
```
that would be via inference SDK

hansent replied:
decided not to do this.  disabled model on hosted lambda

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+
+        predictions = []
+
+        if not boxes:
```
`boxes` would be always set, do u check for batch size here? what is the purpose of this construction?

hansent replied:
currently we have as optional property since SAM2 can predict without prompt...although it's results are much better when prompted, so open to have it be required property instead (also gets around weirdness of how to handle class ids/names when not using object detection boxes as prompts).

PawelPeczek-Roboflow replied:
ok, I got the point and that would almost work as u described, but the field:
```
boxes: StepOutputSelector(
        kind=[
            BATCH_OF_OBJECT_DETECTION_PREDICTION_KIND,
            BATCH_OF_INSTANCE_SEGMENTATION_PREDICTION_KIND,
            BATCH_OF_KEYPOINT_DETECTION_PREDICTION_KIND,
        ]
    ) = Field(  # type: ignore
        description="Boxes (from other model predictions)",
        examples=["$steps.object_detection_model.predictions"],
        default=None,
    )
```
is declared as the value that does not tolerate `None` with `None` as default. Let's fix that, apart from this the approach is fine.

Interestingly, despite my worries about pydantic validation error - pydantic does not complain on manifest init like that:

```python
BlockManifest(
    name="a",
    type="roboflow_core/segment_anything@v1",
    images="$inputs.image",
)
```

probably, defaults are by default excluded from validation based on typing

hansent replied:
Should we make the type `Optional[StepOutputSelector(...)]`?

It seems to work, and in UI it is categorized under optional properties as is

PawelPeczek-Roboflow replied:
yes, you are right

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+            prompt_index = 0
+
+            prompts = Sam2PromptSet()
+            if boxes_for_image is not None:
```
block prevents empty inputs so effectively this if is not needed

hansent replied:
need the check here, because we potentially padded this on line 162 though or you are saying the EE will always provided array with None already and we can scrap 162?

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+
+                    prompt_class_ids.append(boxes_for_image.class_id[prompt_index])
+                    prompt_class_names.append(
+                        boxes_for_image.data["class_name"][prompt_index]
```
use constant for `class_name`

hansent replied:
unsure which one to use or add a new one.

I see:
 - KEYPOINTS_CLASS_NAME_KEY_IN_INFERENCE_RESPONSE here: https://github.com/roboflow/inference/blob/sam2-block/inference/core/workflows/execution_engine/constants.py#L9
 
- DetectionsPropert.CLASS_NAME here:  https://github.com/roboflow/inference/blob/sam2-block/inference/core/workflows/core_steps/common/query_language/entities/enums.py#L30  


Maybe second one (although out of scope for UQL) or add a DETECTIONS_CLASS_NAME in exec_engine.constants?

hansent replied:
added as `DETECTIONS_CLASS_NAME_FIELD` const at top of this file for now

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+            predictions=predictions,
+        )
+
+    def _convert_sam2_segmentation_response_to_inference_instances_seg_response(
```
make this an util function and test

hansent replied:
mhh, it seems similar to other utils function in that it is like conversion...but its very SAM / block result specific so don't think it will/should ever be used anywhere else.  but might be easier to test as non class functionI guess

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+        )
+
+    def _convert_sam2_segmentation_response_to_inference_instances_seg_response(
+        self, sam2_segmentation_predictions, image, prompt_class_ids, prompt_class_names
```
lack of typing

PawelPeczek-Roboflow commented on inference/core/workflows/core_steps/models/foundation/segment_anything/v1.py:
```diff
@@ -0,0 +1,293 @@
+from typing import List, Literal, Optional, Type, Union
```
Q: What is designed usage pattern for the block? To "ground" SAM with bboxes predicted by another model I guess. From what I see the classes of original boxes are to be preserved (if present) or default ones to be added by numbers. The latter approach is probably unstable for sequences processing

probicheaux replied:
>Q: What is designed usage pattern for the block? To "ground" SAM with bboxes predicted by another model I guess.

yes

hansent replied:
I think it's stable because when you don't prompt you will only get one mask, If you prompt with multiple boxes, it will assign class based on the class from the box that was prompt for that mask

PawelPeczek-Roboflow commented on tests/workflows/integration_tests/execution/test_workflow_with_sam2.py:
```diff
@@ -0,0 +1,64 @@
+
+
+@pytest.mark.asyncio
+async def test_clip_workflow_when_minimal_valid_input_provided(
```
This is to be adjusted

PawelPeczek-Roboflow commented on tests/workflows/integration_tests/execution/test_workflow_with_sam2.py:
```diff
@@ -0,0 +1,64 @@
+import numpy as np
```
would be good to have unit tests for manifest parsing

PawelPeczek-Roboflow reviewed (commented):
Ideally we would also have integration tests on SAM predictions in py package? Or maybe I do not understand how we gonna bundle the model

PawelPeczek-Roboflow commented on inference/models/sam2/segment_anything2.py:
```diff
@@ -8,34 +8,38 @@
 
```
is line `self.embedding_cache_keys.pop(0)` correct? I mean the 0 as key to pop

PawelPeczek-Roboflow replied:
`embed_image(...)` lack of typing

PawelPeczek-Roboflow replied:
why `score = score[sorted_ind]` is ignored?

hansent replied:
>is line self.embedding_cache_keys.pop(0) correct? I mean the 0 as key to pop
yeas I think we want FIFO for cache

PawelPeczek-Roboflow replied:
a ok, `embedding_cache_keys` is list not dict 👍

PawelPeczek-Roboflow commented on tests/workflows/integration_tests/execution/test_workflow_with_sam2.py:
```diff
@@ -0,0 +1,64 @@
+    assert set(result[0].keys()) == {
+        "predictions",
+    }, "Expected all declared outputs to be delivered"
+    
```
no qualitative check of prediction

PawelPeczek-Roboflow reviewed (changes requested):
As mentioned in previous comments

PawelPeczek-Roboflow commented:
to run in CI, we should:
* probably enable model by default and disable in envs we do not want it (I believe we install SAM deps in `inference` py package, so would be good for people not to be in need to enable flag just to run a model)
* noticed:
```
WARNING  Error formatting response,        serializers.py:66
                             skipping caching:
                             'Sam2SegmentationPrediction'
                             object has no attribute
                             'confidence'
``` 
while running inference, this is from model monitoring, not sure if relevant

PawelPeczek-Roboflow commented:
this is the result of integration test visualised:
<img width="641" alt="image" src="https://github.com/user-attachments/assets/5a4299a6-35ed-48ae-a022-9b72361c3588">
is that intended?

PawelPeczek-Roboflow commented:
I am in favour of adding that in v2, but an idea would be to enable point-based prompting only only box-based

PawelPeczek-Roboflow commented:
no threshold manipulation option for Sam2 segment request, adding locally for tests

PawelPeczek-Roboflow commented:
there is an error raised in `convert_sam2_segmentation_response_to_inference_instances_seg_response(...)` when SAM2 predicts no masks

PawelPeczek-Roboflow commented:
ok, trying to fix the thing I introduced a bug that may violate the ordering of masks
Let's sync when you are online. I am struggling to understand this:
```
            if len(masks.shape) == 3:
                masks = np.expand_dims(masks, axis=0)
                low_res_logits = np.expand_dims(low_res_logits, axis=0)
                scores = np.expand_dims(scores, axis=0)

            predicted_masks = []
            low_res_masks = []
            for mask, score, low_res_logit in zip(masks, scores, low_res_logits):
                sorted_ind = np.argsort(score)[::-1]
                mask = mask[sorted_ind]
                low_res_logit = low_res_logit[sorted_ind]
                mask = mask[:1]
                low_res_logit = low_res_logit[:1]
                predicted_masks.append(mask)
                low_res_masks.append(low_res_logit)
```
in SAM2 code. Looks like we artificially expand dim but I cannot find a reason why to do it - as a result, we have `(1, C, H, W)` shape for masks and this additional `1` creates issues later on.

hansent commented:
> to run in CI, we should:
> 
> * probably enable model by default and disable in envs we do not want it (I believe we install SAM deps in `inference` py package, so would be good for people not to be in need to enable flag just to run a model)
> * noticed:
> 
> ```
> WARNING  Error formatting response,        serializers.py:66
>                              skipping caching:
>                              'Sam2SegmentationPrediction'
>                              object has no attribute
>                              'confidence'
> ```
> 
> while running inference, this is from model monitoring, not sure if relevant

We need rasterio for polygon -> mask conversion in order for prompt on model to be able to include polygon mask for iterative improvement like smart poly (installed in most dockers via `requirements.sam.txt` already)

probicheaux commented:
>in SAM2 code. Looks like we artificially expand dim but I cannot find a reason why to do it - as a result, we have (1, C, H, W) shape for masks and this additional 1 creates issues later on.

It's not (1, C, H, W), its (1, N, H, W) where N is either 1 or 3 depedning on if multi_mask_output is True or False ---

masks returned shape is really (P, N, H, W) where P is the number of prompts and N is either 1 or 3 masks per prompt. We do the expand dims because in the case were P=1 (one prompt) that dimension is stripped :/ so it's easier for consistency (When we loop through the masks in the downstream code).

I'm curious what downstream problems it's causing?

hansent commented:
> > in SAM2 code. Looks like we artificially expand dim but I cannot find a reason why to do it - as a result, we have (1, C, H, W) shape for masks and this additional 1 creates issues later on.
> 
> It's not (1, C, H, W), its (1, N, H, W) where N is either 1 or 3 depedning on if multi_mask_output is True or False ---
> 
> masks returned shape is really (P, N, H, W) where P is the number of prompts and N is either 1 or 3 masks per prompt. We do the expand dims because in the case were P=1 (one prompt) that dimension is stripped :/ so it's easier for consistency (When we loop through the masks in the downstream code).
> 
> I'm curious what downstream problems it's causing?

Pawel and I paired earlier and confirmed this.  I think its workging as expected.  We adjusted slightly to filter out returned detections in block output when mask was empty (we added threshold parameter so this is now possible)

PawelPeczek-Roboflow commented on tests/inference/models_predictions_tests/conftest.py:
```diff
@@ -157,6 +163,31 @@ def yolov10_det_model() -> Generator[str, None, None]:
+    shutil.rmtree(model_cache_dir)
+
+@pytest.fixture(scope="function")
+def sam2_small_truck_mask() -> Generator[np.ndarray, None, None]:
```
If I understand correctly, this mask come from original SAM2 model, right?

PawelPeczek-Roboflow replied:
it's quite a big file to host in repo - could we assert either low res mask or just a sum of values?

hansent replied:
sum of values was kind of hard to guess good number of allowed error since the elements were unbounded.

Just tried serializing logits and got 257K file.  We could add these to test resources bucket instead (Peter actually suggested this while we were doing it, but I was being lazy and thought it wasn't that much bigger than an image😬), but main issue now would be we need to rewrite history to get large file out of git properly.

PawelPeczek-Roboflow commented on tests/inference/models_predictions_tests/test_sam2.py:
```diff
@@ -0,0 +1,77 @@
+    masks, scores, low_res_logits = model.segment_image(truck_image, prompts=prompt)    
+    
+    
+    #vislualization fo result for debugging
```
lets remove

PawelPeczek-Roboflow commented on tests/inference/models_predictions_tests/test_sam2.py:
```diff
@@ -0,0 +1,77 @@
+import numpy as np
+import pytest
```
let's test also usage of cache for `SegmentAnything2` class

PawelPeczek-Roboflow commented:
@probicheaux your PR with extras was merged here and reverted due to failures of CI. Please fix

hansent commented:
> @hansent Can you add SAM 2 to the docs table of contents in https://github.com/roboflow/inference/blob/main/mkdocs.yml?

is that autogenerated?  looks like its already there, or do I need to add it somewhere else?
>https://github.com/roboflow/inference/blob/main/mkdocs.yml#L56
